### PR TITLE
feat: try to resolve root for new files

### PR DESCRIPTION
**Pull request by mjlbach**
There were issues with fugitive and other non filesystem buffer paths before. Still needs testing for edge cases...

### CODE REVIEW

File: lua/lspconfig/util.lua [hunk 1]
```diff
@@ -161,12 +161,12 @@ M.path = (function()
...
   -- Iterate the path until we find the rootdir.
   local function iterate_parents(path)
-    path = uv.fs_realpath(path)
+    path = uv.fs_realpath(path) or path
     local function it(s, v)
       if not v then
         return
       end
-      if is_fs_root(v) then
+      if is_fs_root(v) or not uv.fs_realpath(path) then
```
Comment (rish987) [inline lua/lspconfig/util.lua:169]: I don't think this works, at least from the perspective of #1045 -- with [Lua generic for](https://www.lua.org/pil/7.2.html), the iterator is called to get the initial value, so in this case we'd be deferring the `nil` return to here and the loop below still wouldn't enter.

File: lua/lspconfig/util.lua [hunk 2]
```diff
@@ -161,12 +161,12 @@ M.path = (function()
 
   -- Iterate the path until we find the rootdir.
   local function iterate_parents(path)
-    path = uv.fs_realpath(path)
+    path = uv.fs_realpath(path) or path
     local function it(s, v)
```
Comment (rish987) [inline lua/lspconfig/util.lua:165]:
```suggestion
    local function it(s, v)
    v = dirname(v)
```

File: lua/lspconfig/util.lua (hunk 1 above)
Comment (rish987) [reply lua/lspconfig/util.lua:169]:
```suggestion
      if is_fs_root(v) or not uv.fs_realpath(v) then
```

File: lua/lspconfig/util.lua [hunk 3]
```diff
@@ -161,12 +161,12 @@ M.path = (function()
...
       if not v then
         return
       end
-      if is_fs_root(v) then
+      if is_fs_root(v) or not uv.fs_realpath(path) then
         return
       end
       return dirname(v), path
```
Comment (rish987) [inline lua/lspconfig/util.lua:172]:
```suggestion
      return v, path
```

**Review by rish987 — COMMENTED** [review]
_I believe_ this is what needs to be done to ensure that we still iterate over all of the parents, even if the given file doesn't exist yet.

File: lua/lspconfig/util.lua [hunk 4]
```diff
@@ -161,12 +161,12 @@ M.path = (function()
 
   -- Iterate the path until we find the rootdir.
   local function iterate_parents(path)
-    path = uv.fs_realpath(path)
+    path = uv.fs_realpath(path) or path
     local function it(s, v)
       if not v then
```
Comment (rish987) [inline lua/lspconfig/util.lua:166]: Also with these changes I'm pretty sure this block can now be removed.

**Comment by mjlbach** [comment]
You can try the latest version @rish987

File: lua/lspconfig/util.lua [hunk 5]
```diff
@@ -161,15 +161,16 @@ M.path = (function()
 
   -- Iterate the path until we find the rootdir.
   local function iterate_parents(path)
-    path = uv.fs_realpath(path)
     local function it(s, v)
-      if not v then
+      if not is_fs_root(v) then
```
Comment (rish987) [inline lua/lspconfig/util.lua:165]:
```suggestion
      if v and not is_fs_root(v) then
```
(to make it consistent w/ previous behavior when `path == nil`)

File: lua/lspconfig/util.lua [hunk 6]
```diff
@@ -263,7 +264,15 @@ function M.search_ancestors(startpath, func)
   if func(startpath) then
     return startpath
   end
+  local guard = 100
   for path in M.path.iterate_parents(startpath) do
+    print(path)
```
Comment (rish987) [inline lua/lspconfig/util.lua:269]:
```suggestion
```
(I think you left this here by accident)

**Comment by rish987** [comment]
Awesome, just tried it and this is working for us now! LGTM

**Comment by mjlbach** [comment]
Great! Thanks for interactively edge-case testing :)
